# HW1: Know the data
HW1 introduces you the dataset which will be used for this and other assignments together with some useful preprocessing methods. Later assignments will cover different classification methods.

In [3]:
import os
import sys
sys.path.append("../../homework")
import matplotlib.pyplot as plt
import numpy as np

%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Fashion Mnist 
Read [here](https://github.com/zalandoresearch/fashion-mnist) to learn more about the Fashion Mnist  dataset. 

In [6]:
# Load the Fashion Mnist  data
import mnist_reader
trainX, trainy = mnist_reader.load_mnist('../data/', kind='train')
testX, testy = mnist_reader.load_mnist('../data/', kind='t10k')


In [7]:
# Print some statistics about the data
print("train image data shape:", trainX.shape)
print("train label data shape:", trainy.shape)
print("test image data shape:", testX.shape)
print("test label data shape:", testy.shape)

train image data shape: (60000, 784)
train label data shape: (60000,)
test image data shape: (10000, 784)
test label data shape: (10000,)


## Visualize the flattened data
The 28x28 Fashion Mnist images have been flattened as 784 x 1 image vectors.

Flattening the data will lead to some loss of information (such as spatial correlation between pixels) but it is nevertheless useful for training some linear classifiers.

Write some code below to visualize the “flattened” data. From the visualization, is “flattened data” sufficiently discriminative for classification purposes? Please provide a brief explanation.

In [ ]:
#####################################################################
# TODO:                                                             #
# Visualize the data to illustrate that the flattened data is still #
# good for classification                                           #
#####################################################################
# your code
#####################################################################
#                       END OF YOUR CODE                            #
#####################################################################

## A few more visualization exercises

Please complete the following to get a better understanding of the Fashion Mnist data:

**a.** Reshape the flattened data back to 2D and plot "average" images of digits 0-9 in one figure. What does this show or represent?

**b.** Subtract the average pixel value computed from all training samples and then re-plot the above “average" images. How are these images useful?

In [ ]:
#####################################################################
# TODO:                                                             #
# Plot each "average image" of digits 0-9                           #
#####################################################################
# your code
#####################################################################
#                       END OF YOUR CODE                            #
#####################################################################

In [ ]:
#####################################################################
# TODO:                                                             #
# Plot the "average" images of digits 1~9 in one figure             #
# (this time with average pixel values subtracted)                  #
#####################################################################
# your code
#####################################################################
#                       END OF YOUR CODE                            #
#####################################################################

In your submission, in addition to the completed codes and the resulting images for the visualizations, also provide brief responses to the above questions and a brief analysis of the visualized data. 

You may refer to [here](http://colah.github.io/posts/2014-10-Visualizing-MNIST/) for more interesting and beautiful visualisations. Of course, the methods employed are more complex.